src/GdbConnection.cc

```diff
@@ -423,17 +423,17 @@ static string read_target_desc(const char* file_name) {
   return ss.str();
 }
 
-static const char* target_description_name(uint32_t cpu_features) {
+static const char* target_description_name(uint32_t cpu_features, bool target_wine) {
   // This doesn't scale, but it's what gdb does...
   switch (cpu_features) {
     case 0:
       return "i386-linux.xml";
     case GdbConnection::CPU_X86_64:
       return "amd64-linux.xml";
     case GdbConnection::CPU_AVX:
-      return "i386-avx-linux.xml";
+      return target_wine ? "i386-avx-wine.xml" : "i386-avx-linux.xml";
```

Comment:
Why do we need special Wine support here?

Reply:
That is how winedbg works currently. It supplies a "custom" xml file, and later executes the "sharedlibrary" command.
If I supply gdb with the regular xml file it uses the "svr4" ops and then it does not "qXfer:libraries" anymore.

That is the backtrace where it requests libraries "winedbg way":
```
(rr) bt
#0  remote_target::remote_read_qxfer (this=0x55f6dbcc09d0, object_name=0x55f6d9ceed86 "libraries", annex=0x0, readbuf=0x55f6dbd3dd60 "`\263>\024+\177", offset=0, len=4096, xfered_len=0x7fff4af67e48, packet=0x55f6d9f5d738 <remote_protocol_packets+504>) at /build/gdb-Nav6Es/gdb-10.1/gdb/remote.c:10881
#1  0x000055f6d9acb1be in target_xfer_partial (ops=0x55f6dbcc09d0, object=TARGET_OBJECT_LIBRARIES, annex=0x0, readbuf=0x55f6dbd3dd60 "`\263>\024+\177", writebuf=0x0, offset=0, len=<optimized out>, xfered_len=0x7fff4af67e48) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1160
#2  0x000055f6d9acda0f in target_read_partial (xfered_len=0x7fff4af67e48, len=4096, offset=0, buf=<optimized out>, annex=0x0, object=TARGET_OBJECT_LIBRARIES, ops=0x55f6dbcc09d0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1387
#3  target_read_alloc_1<char> (ops=0x55f6dbcc09d0, object=object@entry=TARGET_OBJECT_LIBRARIES, annex=annex@entry=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1722
#4  0x000055f6d9accf90 in target_read_stralloc (ops=<optimized out>, object=object@entry=TARGET_OBJECT_LIBRARIES, annex=annex@entry=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1761
#5  0x000055f6d9a75f57 in solib_target_current_sos () at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-target.c:237
#6  0x000055f6d9a7891c in update_solib_list (from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib.c:727
#7  0x000055f6d9a78f07 in solib_add (pattern=0x0, from_tty=1, readsyms=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib.c:961
#8  0x000055f6d985f76c in cmd_func (cmd=0x55f6dbc78e50, args=0x0, from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/cli/cli-decode.c:2181
#9  0x000055f6d9ad7a54 in execute_command (p=<optimized out>, p@entry=<error reading variable: value has been optimized out>, from_tty=1, from_tty@entry=<error reading variable: value has been optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/top.c:668
#10 0x000055f6d99af833 in catch_command_errors (command=<optimized out>, arg=<optimized out>, from_tty=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:457
#11 0x000055f6d99b3147 in captured_main_1 (context=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1218
#12 0x000055f6d99b357b in captured_main (data=data@entry=0x7fff4af68420) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1243
#13 gdb_main (args=args@entry=0x7fff4af68450) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1268
#14 0x000055f6d97cd90c in main (argc=<optimized out>, argv=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/gdb.c:32
(rr) print cmd.name
$1 = 0x55f6d9d17b49 "sharedlibrary"
```


But while collecting this I found following interesting:
```
...remote_target::remote_read_qxfer (..., object_name=0x55f6d9ced563 "memory-map", ...
```
Maybe implementing this could improve situation.


And second, it doesn't "qXfer:libraries" but it looks like it does "qXfer:libraries-svr4" instead:
```
(rr) bt
#0  remote_target::remote_read_qxfer (this=0x562bbc3109d0, object_name=0x562bba9c7af9 "libraries-svr4", annex=0x0, readbuf=0x562bbc38d980 "", offset=0, len=4096, xfered_len=0x7fffe2898ca8, packet=0x562bbac18750 <remote_protocol_packets+528>) at /build/gdb-Nav6Es/gdb-10.1/gdb/remote.c:10881
#1  0x0000562bba7861be in target_xfer_partial (ops=0x562bbc3109d0, object=TARGET_OBJECT_LIBRARIES_SVR4, annex=0x0, readbuf=0x562bbc38d980 "", writebuf=0x0, offset=0, len=<optimized out>, xfered_len=0x7fffe2898ca8) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1160
#2  0x0000562bba788a0f in target_read_partial (xfered_len=0x7fffe2898ca8, len=4096, offset=0, buf=<optimized out>, annex=0x0, object=TARGET_OBJECT_LIBRARIES_SVR4, ops=0x562bbc3109d0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1387
#3  target_read_alloc_1<char> (ops=0x562bbc3109d0, object=object@entry=TARGET_OBJECT_LIBRARIES_SVR4, annex=annex@entry=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1722
#4  0x0000562bba787f90 in target_read_stralloc (ops=<optimized out>, object=object@entry=TARGET_OBJECT_LIBRARIES_SVR4, annex=annex@entry=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1761
#5  0x0000562bba72da95 in svr4_current_sos_via_xfer_libraries (list=0x7fffe2898dd0, annex=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-svr4.c:1235
#6  0x0000562bba72f0e6 in svr4_current_sos_direct (info=0x562bbc30d590) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-svr4.c:1381
#7  0x0000562bba72f7a5 in svr4_current_sos_1 (info=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-svr4.c:1446
#8  svr4_current_sos () at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-svr4.c:1455
#9  0x0000562bba73391c in update_solib_list (from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib.c:727
#10 0x0000562bba733f07 in solib_add (pattern=0x0, from_tty=1, readsyms=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib.c:961
#11 0x0000562bba51a76c in cmd_func (cmd=0x562bbc2c8e50, args=0x0, from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/cli/cli-decode.c:2181
#12 0x0000562bba792a54 in execute_command (p=<optimized out>, p@entry=<error reading variable: value has been optimized out>, from_tty=1, from_tty@entry=<error reading variable: value has been optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/top.c:668
#13 0x0000562bba66a833 in catch_command_errors (command=<optimized out>, arg=<optimized out>, from_tty=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:457
#14 0x0000562bba66e147 in captured_main_1 (context=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1218
#15 0x0000562bba66e57b in captured_main (data=data@entry=0x7fffe28992f0) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1243
#16 gdb_main (args=args@entry=0x7fffe2899320) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1268
#17 0x0000562bba48890c in main (argc=<optimized out>, argv=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/gdb.c:32
(rr) print cmd.name
$2 = 0x562bba9d2b49 "sharedlibrary"
```


I will investigate what those two options reveal ...

Reply:
"memory-map" seems more to describe which memory is used, but not related to libraries. So that is an dead end.

Reply:
Just some more context (at least for me, to make it easier to remember):
Gdb does by default use the `solib_target_so_ops`.
Just when all received target definitions match then gdb starts to use `svr4_so_ops`.
Therefore winedbg seems to not supply a perfect definition on purpose.

That verification takes place here:
```
(rr) bt
#0  0x0000560d3ea8b84c in tdesc_find_register_early (name=name@entry=0x560d3ec4c467 "orig_rax", feature=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/target-descriptions.c:806
#1  tdesc_numbered_register (feature=<optimized out>, data=data@entry=0x560d3f1d2f40, regno=regno@entry=154, name=name@entry=0x560d3ec4c467 "orig_rax") at /build/gdb-Nav6Es/gdb-10.1/gdb/target-descriptions.c:819
#2  0x0000560d3e7d9e3e in amd64_linux_init_abi (info=..., gdbarch=0x560d3f1df530) at /build/gdb-Nav6Es/gdb-10.1/gdb/amd64-linux-tdep.c:1878
#3  0x0000560d3e92fdc2 in i386_gdbarch_init (info=..., arches=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/i386-tdep.c:8705
#4  0x0000560d3e90ed52 in gdbarch_find_by_info (info=...) at /build/gdb-Nav6Es/gdb-10.1/gdb/gdbarch.c:5496
#5  0x0000560d3e7e010e in gdbarch_update_p (info=...) at /build/gdb-Nav6Es/gdb-10.1/gdb/arch-utils.c:538
#6  0x0000560d3ea8aa39 in target_find_description () at /build/gdb-Nav6Es/gdb-10.1/gdb/target-descriptions.c:569
#7  target_find_description () at /build/gdb-Nav6Es/gdb-10.1/gdb/target-descriptions.c:530
#8  0x0000560d3ea39ac3 in remote_target::start_remote (this=this@entry=0x560d3f1709d0, from_tty=from_tty@entry=1, extended_p=extended_p@entry=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/remote.c:4644
#9  0x0000560d3ea3a411 in remote_target::open_1 (name=<optimized out>, from_tty=1, extended_p=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/remote.c:5668
#10 0x0000560d3eaa8cbe in open_target (args=0x7ffcd1a0df53 "127.0.0.1:50868", from_tty=1, command=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:242
#11 0x0000560d3e83b76c in cmd_func (cmd=0x560d3f1115c0, args=0x7ffcd1a0df53 "127.0.0.1:50868", from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/cli/cli-decode.c:2181
#12 0x0000560d3eab3a54 in execute_command (p=<optimized out>, p@entry=<error reading variable: value has been optimized out>, from_tty=1, from_tty@entry=<error reading variable: value has been optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/top.c:668
#13 0x0000560d3e98b833 in catch_command_errors (command=<optimized out>, arg=<optimized out>, from_tty=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:457
#14 0x0000560d3e98f147 in captured_main_1 (context=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1218
#15 0x0000560d3e98f57b in captured_main (data=data@entry=0x7ffcd1a0c840) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1243
#16 gdb_main (args=args@entry=0x7ffcd1a0c870) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1268
#17 0x0000560d3e7a990c in main (argc=<optimized out>, argv=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/gdb.c:32
```

Reply:
As far as I see libraries-svr4 request depends on having some dynamic loader structures, which do not exist.
Therefore the libraries request seems better suiting for this? ( [qXfer:libraries:read](https://sourceware.org/gdb/current/onlinedocs/gdb/Library-List-Format.html) and [qXfer:libraries-svr4:read](https://sourceware.org/gdb/current/onlinedocs/gdb/Library-List-Format-for-SVR4-Targets.html) )